File: LinkedList/Java/DoublyCLL.java


class node
{
  public int data;
  public node next;
  public node prev;
  
  public node()
  {
    data = 0;
    next = null;
    prev = null;
  }

  public node(int no)
  {
    data = no;
    next = null;
    prev = null;
  }
}


class DCLL
{
  private int iCount;
  private node first;
  private node last;

  public DCLL()
  {
    iCount = 0;
    first = null;
    last = null;
  }


  void Display()
  {
    node temp = first;
    do
    {
      System.out.print("| "+temp.data+" |<=>");
      temp = temp.next;
    }while(temp != first);
    System.out.println(" null");
  }
  void InsertFirst(int no)
  {
    node newn = new node(no);
    if(first == null && last == null)
    {
      first = newn;
      last = newn;
      
    }
    else 
    {
      newn.next = first;
      first=newn;
    }
    last.next = first;
    first.prev = last;
    iCount++;
  }

  void InsertLast(int no)
  {
    node newn = new node(no);
    if(first == null && last == null)
    {
      first = newn;
      last = newn;
    }
    else
    {
      last.next = newn ;
      last = newn;
    }
    last.next = first;
    first.prev = last;
    iCount++;
  }

  void InsertAtPos(int no,int pos)
  {
    node newn = new  node(no);
    node temp = null;
    if(pos < 1 || pos > iCount+1)
    {
      System.out.println("Invalid Position");
      return;
    }

    if(pos == 1)
    {
      InsertFirst(no);
    }
    else if (pos == iCount+1)

    {
      InsertLast(no);
    }
    else
    {
      temp = first;
      for(int iCnt = 1 ;iCnt<pos-1;iCnt++)
      {
        temp = temp.next;
      }
      newn.next = temp.next;
      newn.prev = temp;
      temp.next = newn;
      temp.next.prev = newn;
    }
  }

  public void DeleteFirst()
  {
    if(first == null && last == null)
    {
      return;
    }
    else if (first == last)
    {
      first = null;
      last = null;
    }
    else
    {
      first = first.next;
    }
    last.next=first;
    first.prev=last;
    iCount--;
  }

  public void DeleteLast()
  {
    if(first == null && last == null)
    {
      return;
    }
    else if ( first == last)
    {
      first = null;
      last = null;
    }
    else 
    {
      last = last.prev;
    }
    last.next = first;
    first.prev = last;
    iCount--;
  }

  public void DeleteAtPos(int pos)
  {
    node temp = null;
    if(pos < 1 ||pos > iCount)
    {
      System.out.println("Invalid Position");
      return;
    }
    if(pos == 1)
    {
      DeleteFirst();
    }
    else if(pos == iCount)
    {
      DeleteLast();
    }
    else
    {
      temp = first;
      for(int iCnt = 1;iCnt<pos -1;iCnt++)
      {
        temp = temp.next;
      }
      temp.next = temp.next.next;
      temp.next.prev = temp;
      iCount--;
    }
  }
}

public class DoublyCLL {
  public static void main(String A[]) {
      DCLL dcobj = new DCLL();
      dcobj.InsertFirst(21);
      dcobj.InsertFirst(31);
      dcobj.InsertFirst(42);
      dcobj.Display();

      dcobj.InsertLast(55);
      dcobj.InsertLast(88);
      dcobj.InsertLast(11);
      dcobj.Display();
      dcobj.InsertAtPos(78, 3);
      dcobj.Display();
      dcobj.DeleteFirst();
      dcobj.Display();
      dcobj.DeleteAtPos(3);
      dcobj.Display();
  }
}

File: LinkedList/C/DoublyCLL.c
#include<stdio.h>
#include<stdlib.h>


struct node 
{
  int data;
  struct node * next;
  struct node * prev;
};

typedef struct node NODE;
typedef struct node *PNODE;
typedef struct node ** PPNODE;


void Display(PNODE first , PNODE last)
{
  PNODE temp = NULL;
  temp =  first;
  do
  {
    printf("| %d | <=> ",temp ->data);
    temp = temp -> next;
  } while (temp != first);
  printf("NULL\n");

}
int Count(PNODE first , PNODE last)
{
  PNODE temp = NULL;
  int iCount = 0;
  temp = first;
  do
  {
    iCount++;
    temp = temp->next;
  } while (temp != first);
  return iCount;
  
}
void InsertFirst(PPNODE first , PPNODE last , int no)
{

  PNODE newn = (PNODE)malloc(sizeof(NODE));
  newn -> data = no;
  newn -> next = NULL;

  if(*first == NULL && *last == NULL)
  {
    *first = newn;
    *last = newn;
  }
  else
  {
    newn -> next = *first;
    (*first) -> prev = newn;
    *first = newn;
    (*last) -> next = *first;
    (*first)->prev = *last;
  }
}
void InsertLast(PPNODE first , PPNODE last , int no)
{
 
  PNODE newn = (PNODE)malloc(sizeof(NODE));
  newn -> data = no;
  newn -> next = NULL;
  if(*first == NULL && *last == NULL)
  {
    *first = newn;
    *last = newn;
  }
  else
  {
    (*last) ->next = newn;
    newn -> prev = (*last);
    *last = newn;
    (*last)->next = *first ;
    (*first)->prev = *last;
  }
  
  
}
void InsertAtPos(PPNODE first , PPNODE last , int no,int pos)
{
  PNODE newn = (PNODE)malloc(sizeof(NODE));
  newn -> next = NULL;
  newn -> data = no;
  PNODE temp = NULL;
  int iCount = Count(*first,*last);
  if(pos < 1 || pos > iCount+1)
  {
    printf("Invalid Position\n");
    return;
  }

  if(pos == 1)
  {
    InsertFirst(first,last,no);
  }
  else if (pos == iCount+1)
  {
    InsertLast(first,last,no);
  }
  else
  {
    temp = *first;
    for(int iCnt = 1; iCnt < pos -1;iCnt++)
    {
     temp = temp -> next; 
    }
    newn -> next =temp -> next;
    newn ->prev = temp;
    
    temp -> next -> prev = newn;
    temp -> next = newn;
  }
}

void DeleteFirst(PPNODE first , PPNODE last)
{
  if(*first == NULL && *last == NULL)
  {
    return;
  }
  else if (*last == *first)
  {
    free(*first);
    *first = NULL;
    *last = NULL;
  }
  else
  {
    *first = (*first) -> next;
    free((*last) -> next);
    (*last)->next = *first;
    (*first)->prev = *last;
    
  }
}
void DeleteLast(PPNODE first , PPNODE last)
{
  if(*first == NULL && *last == NULL)
  {
    return;
  }
  else if (*last == *first)
  {
    free(*last);
    *last = NULL;
    *first = NULL;
  }
  else
  {
    *last = (*last) -> prev;
    free((*last) -> next);
    (*first )->prev = *last;
    (*last) -> next = (*first);

  }

}
void DeleteAtPos(PPNODE first , PPNODE last , int pos)
{
  int iCount = Count(*first,*last);
  PNODE temp = NULL;
  if(pos < 1 || pos > iCount)
  {
    printf("Invalid Positon \n");
    return;
  }

  if(pos == 1)
  {
    DeleteFirst(first,last);
  }
  else if (pos == iCount)
  {
    DeleteLast(first,last);
  }
  else
  {
    temp = *first;
    for(int iCnt = 1; iCnt < pos -1;iCnt++)
    {
      temp = temp -> next;
    }
    temp->next = temp -> next ->next;
    free(temp ->next -> prev);
    temp ->next ->prev = temp;
  }
}
int main()
{
  PNODE head = NULL;
  PNODE tail = NULL;
  InsertFirst(&head, &tail,105);
  InsertLast(&head, &tail,15);

  InsertFirst(&head, &tail,125);
  InsertLast(&head, &tail,145);
  InsertAtPos(&head,&tail,123,3);
  Display(head,tail);
  // DeleteFirst(&head,&tail);
  // DeleteLast(&head,&tail);
   DeleteAtPos(&head,&tail,3);
  Display(head,tail);
  return 0;
}